### PR TITLE
Add mypy to pytest

### PR DESCRIPTION
**Pull request by stefsmeets**
Hi all, this PR adds mypy to pytest using the `pytest-mypy` extension. I also added `mypy` to `prospector` and set up the config.

### CODE REVIEW

**Comment by stefsmeets** [comment]
Errors to be fixed:

```
(ewatercycle) ~/r/ewatercycle (mypy *) >> mypy
ewatercycle/forcing.py:350: error: "object" not callable
ewatercycle/parametersetdb/__init__.py:42: error: "AbstractConfig" has no attribute "config"
tests/test_forcing.py:7: error: Need type annotation for 'TEST_INPUT_MARRMOT'
tests/test_forcing.py:29: error: Need type annotation for 'TEST_INPUT_LISFLOOD'
tests/test_forcing.py:65: error: Need type annotation for 'TEST_INPUT_HYPE'
tests/test_forcing.py:87: error: Need type annotation for 'TEST_INPUT_WFLOW'
tests/test_forcing.py:123: error: Need type annotation for 'TEST_INPUT_PCRGLOBWB'
tests/test_forcing.py:167: error: Need type annotation for 'EXPECTED_DIFF_MARRMOT'
tests/test_forcing.py:262: error: Need type annotation for 'EXPECTED_DIFF_LISFLOOD'
tests/test_forcing.py:509: error: Need type annotation for 'EXPECTED_DIFF_HYPE'
tests/test_forcing.py:591: error: Need type annotation for 'EXPECTED_DIFF_WFLOW'
tests/test_forcing.py:726: error: Need type annotation for 'EXPECTED_DIFF_PCRGLOBWB'
tests/test_forcing.py:1018: error: Cannot call function of unknown type
tests/test_forcing.py:1023: error: Incompatible types in assignment (expression has type "Tuple[Any, ...]", variable has type "List[Any]")
```

File: ewatercycle/parametersetdb/config.py [hunk 1]
```diff
@@ -36,7 +38,6 @@ def __init__(self, source: str):
         """
         self.source = source
 
-    @abstractmethod
```
Comment (sverhoeven) [inline ewatercycle/parametersetdb/config.py:39]: Without `@abstractmethod `decorator an instance of the class can be made which is unwanted. Why was it removed?

File: ewatercycle/parametersetdb/config.py [hunk 2]
```diff
@@ -27,7 +28,8 @@ class AbstractConfig(ABC):
         config (dict): Content of config as Python dictionary
 
     """
-    @abstractmethod
+    config: Any = None
+
```
Comment (sverhoeven) [inline ewatercycle/parametersetdb/config.py]: Adding config as class variable make Sphinx doc generator give following warning
```
.../ewatercycle/parametersetdb/config.py:docstring of ewatercycle.parametersetdb.config.AbstractConfig.config:1: WARNING: duplicate object description of ewatercycle.parametersetdb.config.AbstractConfig.config, other instance in apidocs/ewatercycle.parametersetdb.config, use :noindex: for one of them
```

The docs warning can be fixed by removing the class docstring and by giving config its own docstring
```suggestion
    config: Any = None
    """Content of config as Python dictionary"""

```

**Review by sverhoeven — CHANGES_REQUESTED** [review]
mypy, pytest and prospector are not showing any typing errors.

When I introduce a type error in ewatercycle/parametersetdb/config.py then mypy, pytest and prospector are giving an expected error. 

The changes however produces a warning during doc generation ( `cd docs && make html`). Can this warning be squashed, see inline suggestion?

File: ewatercycle/parametersetdb/config.py (hunk 1 above)
Comment (stefsmeets) [reply ewatercycle/parametersetdb/config.py:39]: `mypy` complained that you cannot have an `@abstractmethod` on a class that derives from `ABC`. From my understanding, a class either derives from `ABC`, or it uses `@abstractmethod` to mark methods as abstract. But please correct me if I'm wrong.

File: ewatercycle/parametersetdb/config.py [hunk 3]
```diff
@@ -27,7 +28,10 @@ class AbstractConfig(ABC):
         config (dict): Content of config as Python dictionary
 
     """
-    @abstractmethod
+    config: Any = None
+    config: Any = None
```
Comment (sverhoeven) [inline ewatercycle/parametersetdb/config.py]: Oeps the suggestion created a duplicate line
```suggestion
```

Comment (sverhoeven) [reply ewatercycle/parametersetdb/config.py]: Also doc generation still complains can you remove doc string of the class

File: ewatercycle/parametersetdb/config.py (hunk 1 above)
Comment (sverhoeven) [reply ewatercycle/parametersetdb/config.py:39]: For me having both ABC and decorator is needed to get an error on construction of the abstract class. Also mypy does not complain for me.

Comment (stefsmeets) [reply ewatercycle/parametersetdb/config.py:39]: Alright, I will have another look at this 👍

Comment (stefsmeets) [reply ewatercycle/parametersetdb/config.py:39]: Okay, I have reverted the change, but I don't know how to fix these issues:
```
ewatercycle/parametersetdb/__init__.py:55: error: Cannot instantiate abstract class 'AbstractCopier' with abstract attributes '__init__' and 'save'
ewatercycle/parametersetdb/__init__.py:56: error: Cannot instantiate abstract class 'AbstractConfig' with abstract attributes '__init__', 'config' and 'save'
```

**Comment by stefsmeets** [comment]
@sverhoeven Could you also have a look at these issues:

```
ewatercycle/models/abstract.py:40: error: Item "None" of "Optional[Any]" has no attribute "initialize"
ewatercycle/models/abstract.py:44: error: Item "None" of "Optional[Any]" has no attribute "finalize"
ewatercycle/models/abstract.py:49: error: Item "None" of "Optional[Any]" has no attribute "update"
ewatercycle/models/abstract.py:58: error: Item "None" of "Optional[Any]" has no attribute "get_value"
ewatercycle/models/abstract.py:68: error: Item "None" of "Optional[Any]" has no attribute "set_value"
ewatercycle/models/abstract.py:78: error: Item "None" of "Optional[Any]" has no attribute "get_start_time"
ewatercycle/models/abstract.py:83: error: Item "None" of "Optional[Any]" has no attribute "get_end_time"
ewatercycle/models/abstract.py:88: error: Item "None" of "Optional[Any]" has no attribute "get_current_time"
ewatercycle/models/abstract.py:93: error: Item "None" of "Optional[Any]" has no attribute "get_time_units"
ewatercycle/models/abstract.py:98: error: Item "None" of "Optional[Any]" has no attribute "get_time_step"
ewatercycle/models/abstract.py:103: error: Item "None" of "Optional[Any]" has no attribute "get_output_var_names"
```

They seem to be related to:
`self.bmi: Optional[Bmi] = None`, which will cause errors on all the methods if `initialize`(?) is not called first.

**Review by sverhoeven — APPROVED** [review]
If fixed the mypy errors

**Comment by stefsmeets** [comment]
> If fixed the mypy errors

Thanks @sverhoeven! I'm going to quickly merge this before any new updates to master 😅

**Pull request merged**